File: src/lib/model/sync-attention.ts
import { matVec, dot, scale, softmax, flatten } from "./linalg";

/**
 * SyncAttention
 *
 * Cross-attention layer where:
 *   - The QUERY is derived from the sync matrix (N×N pairwise neuron correlations).
 *   - The KEY and VALUE are derived from the input embedding (embedDim-dimensional vector).
 *
 * This answers: "given the current synchronisation state of the network,
 * which part of the input should we focus on?"
 *
 * Architecture (single-head, single key-value pair):
 *
 *   q = W_q  · flatten(syncMatrix)          shape: (queryDim,)
 *   k = W_k  · embedding                    shape: (queryDim,)
 *   v = W_v  · embedding                    shape: (valueDim,)
 *
 *   score  = q · k / sqrt(queryDim)         scalar
 *   weight = softmax([score])               = 1.0 for a single key
 *   ctx    = weight * v                     shape: (valueDim,)
 *
 * With multiple key-value pairs (embeddings stacked as rows of a matrix),
 * each row is projected to its own k_i / v_i and attended jointly.
 * The context is the weighted sum of the v_i vectors.
 *
 * Weights are initialised with a seeded Xavier-style uniform distribution
 * so the layer is deterministic and ready for gradient updates.
 */
export interface SyncAttentionConfig {
  /** Number of neurons N; sync matrix is N×N, flattened to N² for W_q */
  nNeurons: number;
  /** Dimensionality of the input embedding (key/value source) */
  embedDim: number;
  /** Dimensionality of the query and key projections */
  queryDim: number;
  /** Dimensionality of the value projection (= output size) */
  valueDim: number;
}

export class SyncAttention {
  readonly config: SyncAttentionConfig;

  /** W_q: (queryDim × N²) — projects flattened sync matrix → query */
  readonly Wq: number[][];
  /** W_k: (queryDim × embedDim) — projects embedding → key */
  readonly Wk: number[][];
  /** W_v: (valueDim × embedDim) — projects embedding → value */
  readonly Wv: number[][];

  constructor(config: SyncAttentionConfig, seed = 42) {
    this.config = config;
    const { nNeurons, embedDim, queryDim, valueDim } = config;
    const syncFlatDim = nNeurons * nNeurons;

    this.Wq = initWeights(queryDim, syncFlatDim, syncFlatDim, seed);
    this.Wk = initWeights(queryDim, embedDim, embedDim, seed + 1);
    this.Wv = initWeights(valueDim, embedDim, embedDim, seed + 2);
  }

  /**
   * Returns the attention weight for each embedding position given the sync matrix.
   * Weights are in [0, 1] and always sum to exactly 1.0.
   * Useful for interpretability: which input positions does the sync state attend to?
   *
   * @param syncMatrix  N×N pairwise correlation matrix
   * @param embeddings  Sequence of embedding vectors (shape: seqLen × embedDim)
   * @returns           Weight vector of shape (seqLen,)
   */
  computeWeights(syncMatrix: number[][], embeddings: number[][]): number[] {
    const { queryDim } = this.config;
    const syncFlat = flatten(syncMatrix);
    const query = matVec(this.Wq, syncFlat);
    const scale_factor = 1 / Math.sqrt(queryDim);
    const scores = embeddings.map((emb) => dot(query, matVec(this.Wk, emb)) * scale_factor);
    return softmax(scores);
  }

  /**
   * Forward pass.
   *
   * @param syncMatrix  N×N pairwise correlation matrix (each entry in [-1, 1])
   * @param embeddings  One or more embedding vectors; each row is one position.
   *                    Shape: (seqLen × embedDim). Pass a single vector as [[...]] or [...].
   * @returns           Context vector of shape (valueDim,)
   */
  forward(syncMatrix: number[][], embeddings: number[][] | number[]): number[] {
    const { queryDim, valueDim } = this.config;

    // Normalise embeddings to always be number[][]
    const seqs: number[][] = Array.isArray(embeddings[0])
      ? (embeddings as number[][])
      : [(embeddings as number[])];

    // Query from sync matrix (single query for the whole sequence)
    const syncFlat = flatten(syncMatrix);
    const query = matVec(this.Wq, syncFlat); // (queryDim,)

    const scale_factor = 1 / Math.sqrt(queryDim);

    // Compute scores and values for each embedding position
    const scores: number[] = seqs.map((emb) => {
      const key = matVec(this.Wk, emb); // (queryDim,)
      return dot(query, key) * scale_factor;
    });

    const weights = softmax(scores); // (seqLen,)

    // Weighted sum of projected values
    let context: number[] = new Array(valueDim).fill(0);
    for (let i = 0; i < seqs.length; i++) {
      const value = matVec(this.Wv, seqs[i]); // (valueDim,)
      context = context.map((v, j) => v + weights[i] * value[j]);
    }

    return context;
  }
}

// ── Weight initialisation ─────────────────────────────────────────────────────

/**
 * Xavier uniform initialisation: U(-limit, limit) where limit = sqrt(6 / fanIn).
 * Uses a seeded LCG so weights are deterministic across runs.
 */
function initWeights(
  rows: number,
  cols: number,
  fanIn: number,
  seed: number,
): number[][] {
  const limit = Math.sqrt(6 / fanIn);
  let state = seed >>> 0;

  function next(): number {
    // LCG parameters from Numerical Recipes
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
    return state / 0xffffffff; // [0, 1)
  }

  return Array.from({ length: rows }, () =>
    Array.from({ length: cols }, () => (next() * 2 - 1) * limit),
  );
}

File: model/layers/sync_attention.py
"""
model/layers/sync_attention.py

Cross-attention module: query from flattened sync matrix, key/value from embeddings.

The sync matrix encodes pairwise neuron correlations; attending over the embedding
sequence with sync-derived queries lets the model route information based on how
neurons are currently synchronized.
"""

from __future__ import annotations

import torch
import torch.nn as nn
import torch.nn.functional as F


class SyncAttention(nn.Module):
    """Cross-attention: Q from sync matrix, K/V from embeddings.

    Parameters
    ----------
    n_neurons : int
        Number of neurons N. Sync matrix is N×N (flat dim = N²).
    embed_dim : int
        Dimensionality of each neuron embedding.
    query_dim : int
        Projected query/key dimension.
    value_dim : int
        Projected value dimension (output dim per position).
    """

    def __init__(
        self,
        n_neurons: int,
        embed_dim: int,
        query_dim: int,
        value_dim: int,
    ) -> None:
        super().__init__()
        sync_flat = n_neurons * n_neurons
        self.scale = query_dim ** -0.5

        # No bias — matches TS implementation
        self.W_q = nn.Linear(sync_flat, query_dim, bias=False)
        self.W_k = nn.Linear(embed_dim, query_dim, bias=False)
        self.W_v = nn.Linear(embed_dim, value_dim, bias=False)

    def forward(
        self,
        sync_matrix: torch.Tensor,
        embeddings: torch.Tensor,
    ) -> torch.Tensor:
        """Attend over embeddings using sync-derived query.

        Parameters
        ----------
        sync_matrix : torch.Tensor
            Shape (N, N) — pairwise neuron correlation matrix.
        embeddings : torch.Tensor
            Shape (N, embed_dim) — one vector per neuron.

        Returns
        -------
        torch.Tensor
            Shape (N, value_dim) — attended context, one vector per neuron.
        """
        sync_flat = sync_matrix.flatten()  # (N²,)

        query = self.W_q(sync_flat)           # (query_dim,)
        keys = self.W_k(embeddings)           # (N, query_dim)
        values = self.W_v(embeddings)         # (N, value_dim)

        # Scaled dot-product attention: query vs each key → weights over N neurons
        scores = (keys @ query) * self.scale  # (N,)
        weights = F.softmax(scores, dim=0)    # (N,)

        # Weighted sum of values → (value_dim,), tiled to (N, value_dim)
        context = (weights.unsqueeze(-1) * values).sum(dim=0)  # (value_dim,)
        return context.unsqueeze(0).expand(embeddings.shape[0], -1)  # (N, value_dim)

    def compute_weights(
        self,
        sync_matrix: torch.Tensor,
        embeddings: torch.Tensor,
    ) -> torch.Tensor:
        """Return attention weights (N,) without computing attended values.

        Useful for interpretability / logging.
        """
        sync_flat = sync_matrix.flatten()
        query = self.W_q(sync_flat)
        keys = self.W_k(embeddings)
        scores = (keys @ query) * self.scale
        return F.softmax(scores, dim=0)
